third_party/lua-yaml/lyaml.cc

```diff
@@ -682,8 +684,10 @@ static int dump_node(struct lua_yaml_dumper *dumper)
    case MP_STR:
       str = field.sval.data;
       len = field.sval.len;
+      dval = fpconv_strtod(str, &endptr);
+
       if (yaml_is_null(str, len) || yaml_is_bool(str, len, &unused) ||
-          lua_isnumber(dumper->L, -1)) {
+          lua_isnumber(dumper->L, -1) || (isinf(dval) && endptr == str + len)) {
```

Comment:
NB: I've wrote a simple tonumber/strtod testing code while thinking on this patch.

```lua
local ffi = require('ffi')
local tabulate = require('internal.config.utils.tabulate')

ffi.cdef('double strtod(const char *nptr, char **endptr);')

local function convert_lua(s)
    return tonumber(s)
end

local function convert_c(s)
    local nptr = ffi.cast('char *', s)
    local endptr = ffi.new('char *[1]')
    endptr[0] = nptr + s:len()
    local res = ffi.C.strtod(nptr, endptr)
    if endptr[0] ~= nptr + s:len() then
        return nil
    end
    return res
end

local function stringify(x)
    return x == nil and 'nil' or tostring(x)
end

local function run_case(s)
    return {s, stringify(convert_lua(s)), stringify(convert_c(s))}
end

local cases = {
    '1',
    '2e3',
    '0x2p1',
    '0x2p-1',
    '0x2p+1',
    '9e0260492063',
    'abc',
    '077',
    '0b11',
}

local res = {
    {'input', 'tonumber', 'strtod'},
}
for _, case in ipairs(cases) do
    table.insert(res, run_case(case))
end
print(tabulate.encode(res))
```